src/Common/Services/AppInstallationHandler.cs

```diff
@@ -0,0 +1,204 @@
+using Cmf.Foundation.BusinessOrchestration.EntityTypeManagement.OutputObjects;
+using Cmf.LightBusinessObjects.Infrastructure;
+using Cmf.MessageBus.Messages;
+using Cmf.Services.GenericServiceManagement;
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Cmf.CustomerPortal.Sdk.Common.Services
+{
+    public class AppInstallationHandler : IAppInstallationHandler
+    {
+        private readonly ISession _session;
+        private readonly ICustomerPortalClient _customerPortalClient;
+        private bool _isInstallationFinished = false;
+        private bool _hasInstallationFailed = false;
+
+        public AppInstallationHandler(ISession session, ICustomerPortalClient customerPortalClient)
+        {
+            _session = session;
+            _customerPortalClient = customerPortalClient;
+        }
+
+        #region Private Methods
+
+        private void ProcessDeploymentMessage(string subject, MbMessage message)
+        {
+            if (message != null && !string.IsNullOrWhiteSpace(message.Data))
+            {
+                var messageContentFormat = new { Data = string.Empty, DeploymentStatus = (AppInstallationStatus?)AppInstallationStatus.NotInstalled, StepId = string.Empty };
+                var content = JsonConvert.DeserializeAnonymousType(message.Data, messageContentFormat);
+
+                _session.LogInformation(content.Data);
+
+                if (content.DeploymentStatus == AppInstallationStatus.InstallationFailed || content.DeploymentStatus == AppInstallationStatus.InstallationSucceeded)
+                {
+                    if (content.DeploymentStatus == AppInstallationStatus.InstallationFailed)
+                    {
+                        _hasInstallationFailed = true;
+                    }
+                    _isInstallationFinished = true;
+                }
+            }
+            else
+            {
+                _session.LogInformation("Unknown message received");
+            }
+        }
+
+        private async Task ProcessAppInstallation(string appName, CustomerEnvironmentApplicationPackage customerEnvironmentApplicationPackage, string target, DirectoryInfo outputPath)
+        {
+            switch (target)
+            {
+                case "DockerSwarmOnPremisesTarget":
+                case "PortainerV2Target":
+                case "KubernetesOnPremisesTarget":
+                case "OpenShiftOnPremisesTarget":
+                    // get the attachments of the current customer environment application package
+                    GetAttachmentsForEntityInput input = new GetAttachmentsForEntityInput()
+                    {
+                        Entity = customerEnvironmentApplicationPackage
+                    };
+
+                    await Task.Delay(TimeSpan.FromSeconds(1));
```

Comment:
For now, I'm going to leave it as is, since its similar to the Customer Environment logic. Given the timeline and that it may be a hammer, I don't really have the time to test many times for both the environments and the apps.